cmd/dockerd/daemon.go

```diff
@@ -145,6 +145,13 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 		return err
 	}
 
+	conrainerdNamespace, containerdPluginNamespace, err := daemon.RemapContainerdNamespaces(cli.Config)
+	if err != nil {
+		return err
+	}
+	cli.Config.ContainerdNamespace = conrainerdNamespace
+	cli.Config.ContainerdPluginNamespace = containerdPluginNamespace
```

Comment:
I guess this is the part you weren't happy with (modifying the config here)? There's also code to pass options to BuildKit, but I haven't checked if that runs after this code or before https://github.com/moby/moby/blob/ea729aea4a3991e77171c17bfed7b59fe974c513/cmd/dockerd/daemon.go#L423-L424

Also looking at the `dockerd --validate` case (which is run higher up) 🤔

Reply:
The buildkit one comes after, it's called later in this function 

https://github.com/moby/moby/pull/46786/files#diff-eab6997069d2fc9fd3e50c0721e88a363570b9d3ed03facd987373ab1ac9405bR306

Reply:
So it looks like there's an issue in the current code, due to how the daemon config is handled, and we should not depend on `cli.Config` after the daemon is created. I'm wondering (but need to check the flow / order of events) if we should use the new `daemon.Config()` option I added in https://github.com/moby/moby/pull/47187 - which should return the daemon's config after it made changes (set values etc).